tests/test_gf_unpack2.F90

```diff
@@ -4,197 +4,192 @@
+       & char(k6), char(k6), char(k6), char(k6), char(k6),         &
+       & char(k6), char(k6), char(k6), char(k9), char(k9),         &
+       & char(k9), char(k9), char(0), char(0), char(0), char(0),  &
+                                ! section 4
+       & char(0), char(0), char(0), char(71), char(4), char(0),   &
+       & char(0), char(0), char(9), char(1), char(8), char(2),    &
+       & char(0), char(0), char(0), char(0), char(0), char(0),    &
+       & char(0), char(3), char(6), char(9), char(1), char(0),    &
+       & char(0), char(0), char(0), char(0), char(0), char(0),    &
+       & char(0), char(0), char(0),  char(0), char(0), char(0),   &
+       & char(1), char(0), char(0), char(0), char(0), char(0),    &
+       & char(3), char(0), char(0), char(0), char(n2), char(7),   & 
+       & char(n1), char(11), char(15), char(12), char(20),         &
+       & char(10), char(1), char(0), char(0), char(0), char(0),   &
+       & char(1), char(0), char(1), char(0), char(0), char(0),    &
+       & char(12), char(1), char(0), char(0), char(0), char(0),   &
+                                ! section 5
+       & char(0), char(0), char(0), char(47), char(5),  char(0),  &
+       & char(0), char(0), char(k10), char(0), char(2),  char(0), &
+       & char(0), char(0), char(0), char(0), char(0),  char(0),   &
+       & char(1), char(8), char(1), char(1), char(1), char(Nan),  &
+       & char(Nan), char(Nan), char(Nan), char(0), char(0),        &
+       & char(0), char(0), char(0), char(0), char(0), char(0),    &
+       & char(0), char(0), char(0), char(0), char(0), char(0),    &
+       & char(0), char(0), char(0), char(0), char(0), char(0),    &
+                                ! section 6
+       & char(0), char(0), char(0), char(6), char(6),  char(n3),  &
+                                ! section 7
+       & char(0), char(0), char(0), char(12), char(7),  char(1),  &
+       & char(0), char(0), char(0), char(0), char(0),  char(0),   &
+                                ! section 8
+       & "7", "7", "7", "7" /) 
+
+  print *, ''//NEW_LINE('A')//'Calling grib_info ...'
+  call gribinfo(fgrib, fgrib_len, listsec0, listsec1,  &
+       numlocal, numfields, maxvals, ierr)
+  print *, 'Message length: ', listsec0(3)
+  print *, 'Section 2 occurence: ', numlocal
+  print *, 'Section 2 length: ', maxvals(1)
+
+
+  print *, ''//NEW_LINE('A')//'Testging gf_unpack2 ...'
+
+  ! Normal test, should return no error and allocate csec2
+  ! Start byte = 37 => start bit = 296
+  ! Length of section (in octets) is 6
+  iofst = 296
+  call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  if (ierr .ne. 0 .or. iofst .ne. 384 .or. lencsec2 .ne. 6) stop 1
+  deallocate(csec2)
+
+  ! Offset not to section 2 data, should error without allocating csec2
+  iofst = 1
+  call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  if (ierr .ne. 6) stop 2
+
+  ! Offset to section 2 data with already allocated csec2 array
```

Comment:
This seems like an attempt to make the allocate fail on line 53 of gf_unpack2.f. Unfortunately that is not what happens. Instead, fortran happily allocates more memory and just loses the previously allocated memory, unless you kept a separate copy of the pointer. So there is no way to make an allocation fail in testing, Some code will necessarily remain untested as a result.

Reply:
That's what I was attempting here. I also noticed that there is a nullify statement before the allocate statement in gf_unpack2 anyway so there's really no chance the allocate statement could error I don't think.